lib/internal/Magento/Framework/Controller/Result/Json.php

```diff
@@ -59,10 +60,20 @@ public function setJsonData($jsonData)
     }
 
     /**
-     * {@inheritdoc}
+     * @param HttpResponse|ResponseInterface $response
+     * @return $this
      */
     protected function render(ResponseInterface $response)
     {
+        return $this->renderHttpResponse($response);
+    }
+
+    /**
+     * @param HttpResponse $response
+     * @return $this
+     */
+    private function renderHttpResponse(HttpResponse $response)
```

Comment:
Actually the effect would be the same if signature of `render` method was changed to accept `HttpResponse`. In such case there is no need to introduce private method and clients will see correct type-hinting when use `render` method.

Both options lead to backward incompatible changes.